File: frontend/shader/contract/ts/ripple.ts
import { JsonRpcResponse } from "../app/app";

// ============================================================================
// Ripple Configuration
// ============================================================================

export interface RippleConfig {
  rpcUrl: string;
  network: "mainnet" | "testnet";
}

// ============================================================================
// Ripple API Response Types
// ============================================================================

export interface RippleAccountResponse {
  result?: {
    account_data?: {
      Balance?: string;
    };
  };
}

export interface RippleSubmitResponse {
  result?: {
    tx_json?: {
      hash?: string;
    };
    hash?: string;
  };
}

/**
 * Ripple Connector
 */
export class RippleConnector {
  private readonly config: RippleConfig;

  constructor(config: RippleConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Ripple address");
    }

    try {
      const apiUrl =
        this.config.network === "mainnet"
          ? "https://s1.ripple.com:51234"
          : "https://s.altnet.rippletest.net:51234";

      const response = await fetch(apiUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          method: "account_info",
          params: [
            {
              account: address,
              ledger_index: "validated",
            },
          ],
        }),
      });

      if (!response.ok) {
        throw new Error("Failed to fetch Ripple balance");
      }

      const data = (await response.json()) as RippleAccountResponse & {
        error?: { message?: string };
      };
      if (data.error) {
        throw new Error(data.error.message || "RPC error");
      }

      // Return balance in drops (smallest unit, 1 XRP = 1,000,000 drops)
      return data.result?.account_data?.Balance || "0";
    } catch (error) {
      throw new Error(
        `Ripple getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Ripple address");
    }

    const amountNum = Number.parseFloat(amount);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Ripple transactions require:
      // 1. Getting account info and sequence
      // 2. Building payment transaction
      // 3. Signing with private key
      // 4. Submitting to network
      // This is simplified - in production use ripple-lib or xrpl

      const apiUrl =
        this.config.network === "mainnet"
          ? "https://s1.ripple.com:51234"
          : "https://s.altnet.rippletest.net:51234";

      // Get account info
      const accountResponse = await fetch(apiUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          method: "account_info",
          params: [
            {
              account: fromAddress,
              ledger_index: "validated",
            },
          ],
        }),
      });

      if (!accountResponse.ok) {
        throw new Error("Failed to get account info");
      }

      const accountData =
        (await accountResponse.json()) as RippleAccountResponse & {
          error?: { message?: string };
        };
      if (accountData.error) {
        throw new Error(accountData.error.message || "Failed to get account");
      }

      // Submit transaction
      // Note: Transaction must be signed before submission
      // Transaction would include:
      // TransactionType: "Payment",
      // Account: fromAddress,
      // Destination: toAddress,
      // Amount: amountNum.toString(), // Amount in drops
      // Sequence: accountData.result.account_data.Sequence,
      // Fee: "12", // Minimum fee in drops
      const submitResponse = await fetch(apiUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          method: "submit",
          params: [
            {
              tx_blob: "hex_encoded_signed_transaction", // In production, this would be the signed transaction
            },
          ],
        }),
      });

      if (!submitResponse.ok) {
        throw new Error("Failed to submit Ripple transaction");
      }

      const submitData =
        (await submitResponse.json()) as RippleSubmitResponse & {
          error?: { message?: string };
        };
      if (submitData.error) {
        throw new Error(submitData.error.message || "Transaction failed");
      }

      return (
        submitData.result?.tx_json?.hash ||
        submitData.result?.hash ||
        "transaction_hash"
      );
    } catch (error) {
      throw new Error(
        `Ripple sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    return (
      address.length >= 25 && address.length <= 35 && address.startsWith("r")
    );
  }

  /**
   * Bridge tokens from Ripple to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Ripple";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["ripple", "xrp"];
  }
}

File: store/rust/rust/ripple.rs
//! Ripple (XRP) connector

use crate::{BridgeError, BridgeResult, RippleNetwork, JsonRpcRequest, JsonRpcResponse};
use serde::{Deserialize, Serialize};

/// Ripple connector
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RippleConnector {
    rpc_url: String,
    #[allow(dead_code)]
    network: RippleNetwork,
}

impl RippleConnector {
    pub fn new(rpc_url: &str, network: RippleNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Ripple JSON-RPC
        let params = serde_json::json!({
            "account": address,
            "strict": true,
            "ledger_index": "validated"
        });

        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            method: "account_info".to_string(),
            params: serde_json::to_value(params)
                .map_err(|e| BridgeError::SerializationError(format!("Failed to serialize params: {}", e)))?,
            id: 1,
        };

        let client = reqwest::blocking::Client::new();
        let response = client
            .post(&self.rpc_url)
            .json(&request)
            .send()
            .map_err(|e| BridgeError::NetworkError(format!("RPC request failed: {}", e)))?;

        let json_response: JsonRpcResponse = response
            .json()
            .map_err(|e| BridgeError::DeserializationError(format!("Failed to parse response: {}", e)))?;

        if let Some(error) = json_response.error {
            return Err(BridgeError::NetworkError(format!(
                "RPC error {}: {}",
                error.code, error.message
            )));
        }

        let balance = json_response
            .result
            .as_ref()
            .and_then(|r| r.get("account_data"))
            .and_then(|ad| ad.get("Balance"))
            .and_then(|b| b.as_str())
            .and_then(|s| s.parse::<u64>().ok())
            .unwrap_or(0);

        Ok(balance)
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in drops (1 XRP = 1,000,000 drops)
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid Ripple address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Ripple transactions require signing
        // This assumes the RPC endpoint can handle signed transaction submission
        let params = serde_json::json!({
            "TransactionType": "Payment",
            "Account": from_address,
            "Destination": to_address,
            "Amount": amount.to_string()
        });

        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            method: "submit".to_string(),
            params: serde_json::to_value(params)
                .map_err(|e| BridgeError::SerializationError(format!("Failed to serialize params: {}", e)))?,
            id: 1,
        };

        let client = reqwest::blocking::Client::new();
        let response = client
            .post(&self.rpc_url)
            .json(&request)
            .send()
            .map_err(|e| BridgeError::NetworkError(format!("RPC request failed: {}", e)))?;

        let json_response: JsonRpcResponse = response
            .json()
            .map_err(|e| BridgeError::DeserializationError(format!("Failed to parse response: {}", e)))?;

        if let Some(error) = json_response.error {
            return Err(BridgeError::NetworkError(format!(
                "RPC error {}: {}",
                error.code, error.message
            )));
        }

        let hash = json_response
            .result
            .as_ref()
            .and_then(|r| r.get("tx_json"))
            .and_then(|tx| tx.get("hash"))
            .and_then(|h| h.as_str())
            .map(|s| s.to_string())
            .or_else(|| {
                json_response
                    .result
                    .as_ref()
                    .and_then(|r| r.get("hash"))
                    .and_then(|h| h.as_str())
                    .map(|s| s.to_string())
            })
            .ok_or_else(|| BridgeError::TransactionFailed("Failed to get transaction hash".to_string()))?;
        Ok(hash)
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // Ripple addresses are base58 encoded, start with 'r'
        address.len() >= 25 && address.len() <= 35 && address.starts_with('r')
    }
}

